File: leetcode-cn/Python/145.二叉树的后序遍历.py
#
# @lc app=leetcode.cn id=145 lang=python3
#
# [145] 二叉树的后序遍历
#

# @lc code=start
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        ans = []
        stack = []

        pre = None

        while stack or root:
            while root:
                stack.append(root)
                root = root.left

            root = stack.pop()

            # 没有右子树或者右子树已经被访问过
            if not root.right or root.right == pre:
                ans.append(root.val)
                pre = root
                root = None
            else:
                stack.append(root)
                root = root.right

        return ans


# @lc code=end

File: leetcode-cn/C++/145.二叉树的后序遍历.cpp
/*
 * @lc app=leetcode.cn id=145 lang=cpp
 *
 * [145] 二叉树的后序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root)
    {
        stack<TreeNode*> st;
        vector<int> ans;
        TreeNode* prev = nullptr;

        while (root != nullptr || !st.empty()) {
            while (root != nullptr) {
                st.push(root);
                root = root->left;
            }

            root = st.top();
            st.pop();

            if (root->right == nullptr || root->right == prev) {
                ans.push_back(root->val);
                prev = root;
                root = nullptr;
            } else {
                st.push(root);
                root = root->right;
            }
        }

        return ans;
    }
};
// @lc code=end
